lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+      return failure(); // nothing to do
+
+    // Narrow the index value.
+    auto narrowedIndex =
+        narrowValueToArrayWidth(rewriter, op.input(), op.index());
+    rewriter.replaceOpWithNewOp<ArrayGetOp>(op, op.input(), narrowedIndex);
+    return success();
+  }
+};
+
+// Generic pattern to convert parametric result types.
+struct ParametricTypeConversionPattern : public ConversionPattern {
+  ParametricTypeConversionPattern(MLIRContext *ctx,
+                                  TypeConverter &typeConverter,
+                                  ArrayAttr parameters)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*benefit=*/1,
+                          ctx),
+        parameters(parameters) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<Value, 4> convertedOperands;
+    // Update the result types of the operation
+    bool ok = true;
+    rewriter.updateRootInPlace(op, [&]() {
+      // Mutate result types
+      for (auto &it : llvm::enumerate(op->getResultTypes())) {
+        auto res = evaluateParametricType(op->getLoc(), parameters, it.value());
+        ok &= succeeded(res);
+        if (!ok)
+          return;
+        op->getResult(it.index()).setType(res.getValue());
+      }
+    });
+
+    // Note: 'operands' have already been converted with the supplied type
+    // converter to this pattern. Make sure that we materialize this conversion
+    // by updating the operands to op.
+    op->setOperands(operands);
+
+    return success(ok);
+  };
+  ArrayAttr parameters;
+};
+
+struct HWSpecializePass : public sv::HWSpecializeBase<HWSpecializePass> {
+  void runOnOperation() override;
+};
+
+static void populateTypeConversion(Location loc, TypeConverter &typeConverter,
+                                   ArrayAttr parameters) {
+  // Possibly parametric types
+  typeConverter.addConversion([=](hw::IntType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+  typeConverter.addConversion([=](hw::ArrayType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+
+  // Valid target types.
+  typeConverter.addConversion([](mlir::IntegerType type) { return type; });
+}
+
+// Specializes the provided 'base' module into the 'target' module. By doing
+// so, we create a new module which
+// 1. has no parameters
+// 2. has a name composing the name of 'base' as well as the 'parameters'
+// parameters.
+// 3. Has a top-level interface with any parametric types resolved.
+// 4. Any references to module parameters have been replaced with the
+// parameter value.
+static LogicalResult specializeModule(OpBuilder builder, ArrayAttr parameters,
+                                      HWModuleOp source, HWModuleOp &target) {
+  auto *ctx = builder.getContext();
+  // Update the types of the source module ports based on evaluating any
```

Comment:
I know _nothing_ about the TypeConverter system, but this feels relatively common so there may be something in there that handles this...?

Reply:
There's something in there for "FunctionLike" ops, but that concept has evolved since the last time I tried using it on HW things: https://github.com/llvm/llvm-project/blob/94244bda9d7f2ad49d75a412e4f7f62ab228a544/mlir/lib/Transforms/Utils/DialectConversion.cpp#L3063-L3104. It may work... 

IIRC I tried to use it here, but that didn't quite work for unrelated reasons (impedance mismatch between HW and LLHD modelling): https://github.com/llvm/circt/blob/230c1955f4af25c151989c13e7712d06c158f9f7/lib/Conversion/HWToLLHD/HWToLLHD.cpp#L108. So again, it may work 🤷

Reply:
The tricky thing here is that I'm not updating/rewriting the existing module; I'm creating a new set of ports that are used to create a new module. Signature updating of FunctionLike ops in the dialect conversion concerns itself with in-place modification of existing operations.